### ISSUE TITLE
Removal Note in `parametrizations.spectral_norm`

### ISSUE THREAD

**Issue by GloryyrolG** [pytorch/pytorch#68421]
## 📚 Documentation

In https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrizations.spectral_norm.html, there is a note saying

> If the _SpectralNorm module, i.e., module.parametrization.weight[idx], is in training mode on removal, it will perform another power iteration. If you’d like to avoid this iteration, set the module to eval mode before its removal.

However, I think, [`_SpectralNorm`](https://pytorch.org/docs/1.10.0/_modules/torch/nn/utils/parametrizations.html#_SpectralNorm) does not deal with `remove` with a more forward as [`SpectralNorm`](https://pytorch.org/docs/1.10.0/_modules/torch/nn/utils/spectral_norm.html#SpectralNorm) in `torch.nn.utils.spectral_norm` does. I.e., currently, `parametrizations.spectral_norm` does not support removal and it is better to remove the note from the page.

cc @brianjo @mruberry @albanD @jbschlosser @walterddr

**Comment by jbschlosser** [comment]
cc @Lezcano what are your thoughts on this?

**Comment by Lezcano** [comment]
@GloryyrolG what do you mean with "it does nto support removal"? It may be removed with https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrize.remove_parametrizations.html

**Comment by GloryyrolG** [comment]
Basically, like what is stated in this discussion post https://discuss.pytorch.org/t/how-to-remove-spectralnorm-parametrizations-in-pytorch/137031/4?u=glory_chen

`_SpectralNorm` added in `parametrizations` in the reimplementation does not have a similar `remove` function which does one more time of forward to compute weight.

https://github.com/pytorch/pytorch/blob/53bfb00ee1bef9eedc320aebc10e21fdbb29fef0/torch/nn/utils/parametrizations.py#L485

I.e., it can be removed but will encounter similar problems https://github.com/pytorch/pytorch/issues/51800

**Comment by Lezcano** [comment]
Could you please try the suggestion given above? Please report whether you find any problems with the proposed approach.

**Comment by GloryyrolG** [comment]
## To Reproduce

```python
from torch import nn
import torch

from torch.nn.utils.parametrizations import spectral_norm
from torch.nn.utils.parametrize import remove_parametrizations


class SpectralNormTestModel(nn.Module):
        def __init__(self):
            super(SpectralNormTestModel, self).__init__()
            feature_count = 1000

            def init_(layer: nn.Linear):
                nn.init.orthogonal_(layer.weight.data, 1.372)
                nn.init.zeros_(layer.bias.data)

            self.fc1 = nn.Linear(feature_count, feature_count)
            init_(self.fc1)
            self.fc1 = spectral_norm(self.fc1)
            self.act1 = nn.PReLU(feature_count)

            self.fc2 = nn.Linear(feature_count, feature_count)
            init_(self.fc2)
            self.fc2 = spectral_norm(self.fc2)
            self.act2 = nn.PReLU(feature_count)

            self.out = nn.Linear(feature_count, 1)
            init_(self.out)

        def forward(self, x: torch.Tensor) -> torch.Tensor:
            return self.out(self.act2(self.fc2(self.act1(self.fc1(x)))))
     
def spectral_norm_init_test(use_forward: bool, remove: bool):
    x = torch.randn(10000, 1000, device='cuda')
    model = SpectralNormTestModel().cuda()
    if use_forward:
        y = model.forward(x)  # first forward once

    elif remove:
        # model.eval()
        # remove_spectral_norm(model.fc1)
        remove_parametrizations(model.fc1, 'weight', leave_parametrized=False)
        remove_parametrizations(model.fc2, 'weight', leave_parametrized=False)
    y = model.eval()(x) 
    print(f"mean absolute output: {y.abs().mean()}")

print('----- test with use_forward ------')
spectral_norm_init_test(True, False)
print('----- test without use_forward ------')
spectral_norm_init_test(False, False)
print('----- test after removing sn ------')
spectral_norm_init_test(False, True)
print('----- done ------')

```

## Expected behavior

```bash
----- test with use_forward ------
mean absolute output: 0.5045999884605408
----- test without use_forward ------
mean absolute output: 0.518002986907959
----- test after removing sn ------
mean absolute output: 0.9170235991477966
----- done ------

```

By just modifying a bit the code in [#Issue 51800](https://github.com/pytorch/pytorch/issues/51800#issuecomment-793941555), we can see, this way of removal seems not to work as expected, i.e., it does not compute a more time of [`compute_weight`](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/spectral_norm.py#L43) like that in the normal [`SpectralNorm`](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/spectral_norm.py#L10). Actually, I would expect all the outputs to be similar to each other.

**Comment by Lezcano** [comment]
Removing the randomness of your answer, one gets similar results, as expected.
Note that you were:
- Sampling a new `x` vector in every test
- Sampling a new module in every test.
- When removing the parametrisation, you were not leaving the parameter with the parametrised weight.

Note that, in the test below, there are some minimal differences between the three outputs, but this comes from the nondeterminism of some operations in cuda. If you run this example on CPU the results from the first and third executions are equal, as claimed in the docs.

Closing for now.

```python
from copy import deepcopy
from torch import nn
import torch

from torch.nn.utils.parametrizations import _SpectralNorm
from torch.nn.utils.parametrize import remove_parametrizations, register_parametrization


class SpectralNormTestModel(nn.Module):
    def __init__(self):
        super(SpectralNormTestModel, self).__init__()
        feature_count = 1000

        def init_(layer: nn.Linear):
            nn.init.orthogonal_(layer.weight.data, 1.372)
            nn.init.zeros_(layer.bias.data)

        self.fc1 = nn.Linear(feature_count, feature_count)
        init_(self.fc1)
        self.act1 = nn.PReLU(feature_count)

        self.fc2 = nn.Linear(feature_count, feature_count)
        init_(self.fc2)
        self.act2 = nn.PReLU(feature_count)

        self.out = nn.Linear(feature_count, 1)
        init_(self.out)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.out(self.act2(self.fc2(self.act1(self.fc1(x)))))


def spectral_norm_init_test(use_forward: bool, remove: bool, x, model):
    if use_forward:
        y = model.forward(x)  # first forward once

    elif remove:
        # model.eval()
        # remove_spectral_norm(model.fc1)
        remove_parametrizations(model.fc1, 'weight', leave_parametrized=True)
        remove_parametrizations(model.fc2, 'weight', leave_parametrized=True)
    y = model.eval()(x)
    print(f"mean absolute output: {y.abs().mean()}")


def clone_parametrized_model(model):
    cloned_model = deepcopy(model)
    register_parametrization(cloned_model.fc1, "weight", _SpectralNorm(cloned_model.fc1.weight).cuda())
    register_parametrization(cloned_model.fc2, "weight", _SpectralNorm(cloned_model.fc2.weight).cuda())
    return cloned_model



x = torch.randn(10000, 1000, device='cuda')
model = SpectralNormTestModel().cuda()
model1 = clone_parametrized_model(model)
model2 = clone_parametrized_model(model)
model3 = clone_parametrized_model(model)


print('----- test with use_forward ------')
spectral_norm_init_test(True, False, x.detach().clone(), model1)
print('----- test without use_forward ------')
spectral_norm_init_test(False, False, x.detach().clone(), model2)
print('----- test after removing sn ------')
spectral_norm_init_test(False, True, x.detach().clone(), model3)
print('----- done ------')
```
